File: graph_algorithms/week1_undirected_graphs/1_finding_exit_from_maze/reachability.py
from typing import List


def dfs(adj: List[List[int]], visited: List[bool], u: int, v: int) -> int:
    """Returns 1 if there is a path between u and v, 0 otherwise."""
    if u == v or v in adj[u]:
        return 1

    visited[u] = True

    return int(any(dfs(adj, visited, w, v) for w in adj[u] if not visited[w]))


def reach(adj: List[List[int]], u: int, v: int) -> int:
    """Returns 1 if there is a path between u and v, 0 otherwise."""
    visited = [False] * len(adj)
    return dfs(adj, visited, u, v)


if __name__ == "__main__":
    n, m = map(int, input().split())

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    for a, b in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)

    u, v = map(int, input().split())
    print(reach(adj, u - 1, v - 1))

File: graph_algorithms/week1_undirected_graphs/1_finding_exit_from_maze/reachability.cpp
#include <iostream>
#include <vector>

using std::vector;

// Returns 1 if there is a path between u and v, 0 otherwise.
int dfs(vector<vector<int>> &adj, vector<bool> &visited, int u, int v) {
  if (u == v)
    return 1;

  visited[u] = true;

  for (int i = 0; i < adj[u].size(); i++) {
    int w = adj[u][i];

    if (!visited[w] && dfs(adj, visited, w, v))
      return 1;
  }

  return 0;
}

// Returns 1 if there is a path between u and v, 0 otherwise.
int reach(vector<vector<int>> &adj, int u, int v) {
  vector<bool> visited(adj.size(), false);
  return dfs(adj, visited, u, v);
}

int main() {
  size_t n, m;
  std::cin >> n >> m;

  vector<vector<int>> adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int a, b;
    std::cin >> a >> b;
    adj[a - 1].push_back(b - 1);
    adj[b - 1].push_back(a - 1);
  }

  int u, v;
  std::cin >> u >> v;
  std::cout << reach(adj, u - 1, v - 1) << std::endl;
}
